DataFlash: move starting of logs into DataFlash

peterbarker opened the pull request.

OXINARF commented on libraries/DataFlash/DataFlash_Backend.cpp:
```diff
@@ -260,6 +260,31 @@ bool DataFlash_Backend::Log_Write(const uint8_t msg_type, va_list arg_list, bool
+    if (_front.in_log_download()) {
+        return false;
+    }
+    if (!DataFlash_Backend::WritesOK()) {
```
Any reason for this to be forcing the DataFlash_Backend WritesOK method instead of a possible overriding method?

peterbarker replied:
Yes, the File backend says "writes are not OK" if the file is not open - we're currently checking to see if it is OK to open that file :-)

OXINARF replied:
Humm, makes sense. Although I have to say I dislike that the same method have two slightly different meanings whether it is in base class or in derived class. I wonder if the additional checks in the derived `WritesOK` shouldn't be a separate `CanWrite` method (or a better name). Then in `WritePrioritisedBlock` we could do:

```
if(!WritesOK()) { // we shouldn't write
  return false;
}

if(StartNewLogOK()) { // we need to start a new log
  start_new_log();
}

if(!CanWrite()) { // we should be logging but there was a problem and we can't write
  return false;
}
```

Maybe I'm nitpicking here, but I'm not very comfortable with the way it is now.

peterbarker replied:
I was thinking of splitting the code out of WritesOK into a separate method which was called by WritesOK and by StartNewLogOK.  I couldn't come up with a name I liked.  So I just called the other one.  I'll come up with a name.  It's be non-virtual and private, so it doesn't matter *too* much....

OXINARF commented on libraries/AP_Arming/AP_Arming.cpp:
```diff
@@ -493,8 +493,16 @@ bool AP_Arming::arm_checks(uint8_t method)
-        if (!DataFlash_Class::instance()->logging_started()) {
+        // allow logging to start by indicating (falsely) that the vehicle is armed:
+        DataFlash_Class *df = DataFlash_Class::instance();
+        df->set_vehicle_armed(true);
```
I understand the reason but I don't really like this. What if another arm check fails? It will continue writing to log.

peterbarker replied:
I'm not its greatest fan either!

This is currently supposed to be the *last* arm check - at the moment I don't think we have another arm check which changes system state.  So if the logging succeeds, we will arm.

Logging is a particular problem in that if I double-down on lying to dataflash and tell it we're now disarmed - we may rotate the log file (`LOG_FILE_DSRMROT`) :-)  So I think it may always need to be the last arming check.

I guess we could add post-arm checks, but that's a whole new can of worms.

I did notice, however, that we don't seem to make this check for Copter.   Adding a patch for that.

OXINARF replied:
I agree that disarming after isn't a good idea, I wasn't suggesting that. I don't have a better way (yet), but I really think this isn't a great idea - your finding with Copter is what I'm afraid about, someone will break this some day. By the way, unless I'm missing something, Sub also doesn't go through here.

peterbarker replied:
A couple of ideas that did come to me:
 - post-arming checks which would activate before we set any notify flags
   - you now have race conditions against other threads which may or may not care about your arming state
 - lists of cleanup handlers in the case an arming check fails.
 - another entry point to DataFlash to say "do everything you would do if vehicle was just set to armed (I *really* don't like this option...)
 - restructure to have a `goto BAD_LOGGING:` set of cleanup handlers

I didn't act on any of those, obviously.

peterbarker replied:
Sub's checks are another thing to look at... later...

OXINARF commented:
@peterbarker Sorry if I'm nitpicking too much on this. I like the idea of the PR, I'm just not very comfortable with the implementation.

peterbarker commented:
I added that method as described.  Kinda fixes that structural issue.  Not sure its worth it, 'though.

OXINARF commented:
I think that you didn't really got what my suggestion was, so I've gone ahead and put it in code: https://github.com/OXINARF/ardupilot/commit/e4413a9fefd4e6f5ae47278e885b9d18aafe952d This is with your last commit removed, which I'm not sure I understand - to me it looks like it keeps things exactly the same.

I've also done something for the arming, but I'm not completely convinced by it too, specially because of the arm/disarm stuff: https://github.com/OXINARF/ardupilot/commit/19784cc90cadb7153ee3f01ce10d571160120788 Is there any reason why we shouldn't keep the *StartUnstartedLogging* method which would only be called on this arming check?

peterbarker commented:
On Tue, 4 Jul 2017, Francisco Ferreira wrote:

> I think that you didn't really got what my suggestion was, so I've gone
> ahead and put it in code: OXINARF/ardupilot@e4413a9 This is with your last


*now* I understand, and yes, that's a much nicer way of doing it!

If you want to put that in on top of those patches, please go ahead.

peterbarker commented:
On Tue, 4 Jul 2017, Francisco Ferreira wrote:

> I've also done something for the arming, but I'm not completely 
> convinced by it too, specially because of the arm/disarm stuff: 
> OXINARF/ardupilot@19784cc Is there any reason why we shouldn't keep the

I'm not a great fan of that method.  It's special-casing the logging start 
- there may in future we arming checks that have other side-effects that 
need to be reversed, and that will be difficult if we split it like this.

I really think commenting is sufficient in this case.  Lots of comments. 
I've pushed up a couple of patches which try to add signs indicating the 
pitfalls.

> StartUnstartedLogging method which would only be called on this arming 
> check?

These changes are designed to remove knowledge about the logging system 
from the vehicle code.  StartUnstartedLogging is bad not just because the 
vehicle code magically needs to know a call is needed in a particular 
place - and that's not even the case with DataFlashOverMAVLink :-)

OXINARF commented:
> I'm not a great fan of that method.  It's special-casing the logging start - there may in future we arming checks that have other side-effects that need to be reversed, and that will be difficult if we split it like this.

OK, I agree with it and I can live with just having the comments.

> These changes are designed to remove knowledge about the logging system from the vehicle code. StartUnstartedLogging is bad not just because the vehicle code magically needs to know a call is needed in a particular place - and that's not even the case with DataFlashOverMAVLink :-)

I understand the intention of these changes, but I think it accomplishes the opposite. Let's see, what is this arming check supposed to be doing? Checking that logging has started, right? To do that these changes are telling the dataflash the vehicle is armed (which isn't exactly true) and writing something to force the start of the log - I think this uses much more inside info from DataFlash than having a method telling it to start.

I should mention, in case I wasn't clear, I'm not proposing to throw this patch away and have vehicles start the log like it is now; what I'm proposing is the arming check does what it is supposed to: if logging not started yet, start it and then check if it has started. If you aren't still clear of what I'm proposing, let me know and I'll put it in code.

peterbarker commented:
@OXINARF OK, new patch added which I think is pretty much what you're saying, and seems much cleaner that what I had!

peterbarker commented:
@OXINARF ping on this one; it affects work Randy and Pierre are doing in Rover.

OXINARF commented:
@peterbarker Thanks for the ping. I did have this on my list, but was postponing it since I didn't want to ask for changes again - sorry for being a pain in this one!

After all the back and forth I thought that factoring out the logging check in AP_Arming wasn't bringing anything useful, so I removed that part. I also changed the comment there, so it is more in-line with what the last version of code does. Here is a commit that shows all the changes I did compared to this PR: https://github.com/OXINARF/ardupilot/commit/4a9dcbb883b14cecd8a81562af44daac8c49bad7
Every change in that commit will have you as author, so please check you agree with it, specially the language in the code comment.

Besides that, I squashed, changed commit messages and re-ordered commits, so please take a good look at https://github.com/ArduPilot/ardupilot/compare/master...OXINARF:review/6526  This is what I'd like to merge - almost all commits have you as author so you should see if you agree with it, specially the commit messages.

peterbarker commented:
On Tue, 18 Jul 2017, Francisco Ferreira wrote:

> @peterbarker Thanks for the ping. I did have this on my list, but was postponing it since I didn't want to ask for changes again - sorry for being a pain in this one!

:-)  No worries.

> After all the back and forth I thought that factoring out the logging check in AP_Arming wasn't bringing anything useful, so I removed that part. I also changed the comment there, so it is more in-line with what the last version of code does. Here is a commit that shows all the changes I did compared to this PR:
> OXINARF/ardupilot@4a9dcbb
> Every change in that commit will have you as author, so please check you agree with it, specially the language in the code comment.

Yep, I'm good with that.

> Besides that, I squashed, changed commit messages and re-ordered commits, so please take a good look at master...OXINARF:review/6526 This is what I'd like to merge - almost all commits have you as author so you should see if you agree with it, specially the commit messages.

All good!

OXINARF commented:
Merged, thanks!
